File: src-tauri/src/commands/resource_monitor.rs
use crate::error::CommandError;
use crate::models::{
    CollectResourceBatchInput, CollectResourceBatchResult, ListResourceAlertEventsInput,
    ListResourceAlertRulesInput, ResourceAlertEvent, ResourceAlertRule, ResourceMetricSnapshot,
    ResourceMonitorOverview, ResourceMonitorTarget, ResourceSnapshotListInput,
    UpsertResourceAlertRuleInput, UpsertResourceMonitorTargetInput,
};
use crate::services::resource_monitor::ResourceMonitorService;
use crate::state::AppState;

#[tauri::command]
pub fn list_resource_monitor_targets(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<ResourceMonitorTarget>, CommandError> {
    ResourceMonitorService::list_targets(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_resource_monitor_target(
    state: tauri::State<'_, AppState>,
    input: UpsertResourceMonitorTargetInput,
) -> Result<ResourceMonitorTarget, CommandError> {
    ResourceMonitorService::upsert_target(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_resource_monitor_target(
    state: tauri::State<'_, AppState>,
    target_type: String,
    target_key: String,
) -> Result<(), CommandError> {
    ResourceMonitorService::delete_target(&state.db, &target_type, &target_key)
        .map_err(|e| e.into())
}

#[tauri::command]
pub fn get_resource_monitor_overview(
    state: tauri::State<'_, AppState>,
) -> Result<ResourceMonitorOverview, CommandError> {
    ResourceMonitorService::overview(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn list_resource_metric_snapshots(
    state: tauri::State<'_, AppState>,
    input: ResourceSnapshotListInput,
) -> Result<Vec<ResourceMetricSnapshot>, CommandError> {
    ResourceMonitorService::list_snapshots(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub async fn collect_server_resource_snapshot(
    state: tauri::State<'_, AppState>,
    alias: String,
) -> Result<ResourceMetricSnapshot, CommandError> {
    ResourceMonitorService::collect_server(&state.db, &alias)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn collect_database_resource_snapshot(
    state: tauri::State<'_, AppState>,
    connection_key: String,
) -> Result<ResourceMetricSnapshot, CommandError> {
    ResourceMonitorService::collect_database(&state.db, &connection_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn collect_redis_resource_snapshot(
    state: tauri::State<'_, AppState>,
    connection_key: String,
) -> Result<ResourceMetricSnapshot, CommandError> {
    ResourceMonitorService::collect_redis(&state.db, &connection_key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn collect_resource_snapshots_batch(
    state: tauri::State<'_, AppState>,
    input: CollectResourceBatchInput,
) -> Result<CollectResourceBatchResult, CommandError> {
    ResourceMonitorService::collect_batch(&state.db, input)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub fn list_resource_alert_rules(
    state: tauri::State<'_, AppState>,
    input: ListResourceAlertRulesInput,
) -> Result<Vec<ResourceAlertRule>, CommandError> {
    ResourceMonitorService::list_alert_rules(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_resource_alert_rule(
    state: tauri::State<'_, AppState>,
    input: UpsertResourceAlertRuleInput,
) -> Result<ResourceAlertRule, CommandError> {
    ResourceMonitorService::upsert_alert_rule(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_resource_alert_rule(
    state: tauri::State<'_, AppState>,
    id: i64,
) -> Result<(), CommandError> {
    ResourceMonitorService::delete_alert_rule(&state.db, id).map_err(|e| e.into())
}

#[tauri::command]
pub fn list_resource_alert_events(
    state: tauri::State<'_, AppState>,
    input: ListResourceAlertEventsInput,
) -> Result<Vec<ResourceAlertEvent>, CommandError> {
    ResourceMonitorService::list_alert_events(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn resolve_resource_alert_event(
    state: tauri::State<'_, AppState>,
    id: i64,
) -> Result<(), CommandError> {
    ResourceMonitorService::resolve_alert_event(&state.db, id).map_err(|e| e.into())
}

File: src/lib/api/resourceMonitor.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  CollectResourceBatchInput,
  CollectResourceBatchResult,
  KillMysqlQueryInput,
  KillMysqlQueryResult,
  ListResourceAlertEventsInput,
  ListResourceAlertRulesInput,
  MysqlSlowQuery,
  MysqlSlowQueryListInput,
  ResourceAlertEvent,
  ResourceAlertRule,
  ResourceMetricSnapshot,
  ResourceMonitorOverview,
  ResourceMonitorTarget,
  ResourceSnapshotListInput,
  UpsertResourceAlertRuleInput,
  UpsertResourceMonitorTargetInput,
} from "@/types";

export const resourceMonitorApi = {
  listTargets: () =>
    hasTauriRuntime()
      ? invoke<ResourceMonitorTarget[]>("list_resource_monitor_targets")
      : devApiFetch<ResourceMonitorTarget[]>("/resource-monitor/targets"),
  upsertTarget: (input: UpsertResourceMonitorTargetInput) =>
    hasTauriRuntime()
      ? invoke<ResourceMonitorTarget>("upsert_resource_monitor_target", {
          input,
        })
      : devApiFetch<ResourceMonitorTarget>("/resource-monitor/targets", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  deleteTarget: (targetType: string, targetKey: string) =>
    hasTauriRuntime()
      ? invoke<void>("delete_resource_monitor_target", {
          targetType,
          targetKey,
        })
      : devApiFetch<void>(
          `/resource-monitor/targets/${encodeURIComponent(targetType)}/${encodeURIComponent(targetKey)}`,
          { method: "DELETE" },
        ),
  overview: () =>
    hasTauriRuntime()
      ? invoke<ResourceMonitorOverview>("get_resource_monitor_overview")
      : devApiFetch<ResourceMonitorOverview>("/resource-monitor/overview"),
  listSnapshots: (input: ResourceSnapshotListInput) =>
    hasTauriRuntime()
      ? invoke<ResourceMetricSnapshot[]>("list_resource_metric_snapshots", {
          input,
        })
      : devApiFetch<ResourceMetricSnapshot[]>("/resource-monitor/snapshots", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  collectServer: (alias: string) =>
    hasTauriRuntime()
      ? invoke<ResourceMetricSnapshot>("collect_server_resource_snapshot", {
          alias,
        })
      : devApiFetch<ResourceMetricSnapshot>(
          `/resource-monitor/server/${encodeURIComponent(alias)}/collect`,
          { method: "POST" },
        ),
  collectDatabase: (connectionKey: string) =>
    hasTauriRuntime()
      ? invoke<ResourceMetricSnapshot>("collect_database_resource_snapshot", {
          connectionKey,
        })
      : devApiFetch<ResourceMetricSnapshot>(
          `/resource-monitor/database/${encodeURIComponent(connectionKey)}/collect`,
          { method: "POST" },
        ),
  collectRedis: (connectionKey: string) =>
    hasTauriRuntime()
      ? invoke<ResourceMetricSnapshot>("collect_redis_resource_snapshot", {
          connectionKey,
        })
      : devApiFetch<ResourceMetricSnapshot>(
          `/resource-monitor/redis/${encodeURIComponent(connectionKey)}/collect`,
          { method: "POST" },
        ),
  collectBatch: (input: CollectResourceBatchInput) =>
    hasTauriRuntime()
      ? invoke<CollectResourceBatchResult>("collect_resource_snapshots_batch", {
          input,
        })
      : devApiFetch<CollectResourceBatchResult>(
          "/resource-monitor/collect-batch",
          {
            method: "POST",
            body: JSON.stringify(input),
          },
        ),
  listMysqlSlowQueries: (input: MysqlSlowQueryListInput) =>
    hasTauriRuntime()
      ? invoke<MysqlSlowQuery[]>("list_mysql_slow_queries", { input })
      : devApiFetch<MysqlSlowQuery[]>("/resource-monitor/mysql/slow-queries", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  killMysqlQuery: (input: KillMysqlQueryInput) =>
    hasTauriRuntime()
      ? invoke<KillMysqlQueryResult>("kill_mysql_query", { input })
      : devApiFetch<KillMysqlQueryResult>(
          "/resource-monitor/mysql/kill-query",
          {
            method: "POST",
            body: JSON.stringify(input),
          },
        ),
  listAlertRules: (input: ListResourceAlertRulesInput) =>
    hasTauriRuntime()
      ? invoke<ResourceAlertRule[]>("list_resource_alert_rules", { input })
      : devApiFetch<ResourceAlertRule[]>("/resource-monitor/alert-rules/list", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  upsertAlertRule: (input: UpsertResourceAlertRuleInput) =>
    hasTauriRuntime()
      ? invoke<ResourceAlertRule>("upsert_resource_alert_rule", { input })
      : devApiFetch<ResourceAlertRule>("/resource-monitor/alert-rules", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  deleteAlertRule: (id: number) =>
    hasTauriRuntime()
      ? invoke<void>("delete_resource_alert_rule", { id })
      : devApiFetch<void>(
          `/resource-monitor/alert-rules/${encodeURIComponent(id)}`,
          {
            method: "DELETE",
          },
        ),
  listAlertEvents: (input: ListResourceAlertEventsInput) =>
    hasTauriRuntime()
      ? invoke<ResourceAlertEvent[]>("list_resource_alert_events", { input })
      : devApiFetch<ResourceAlertEvent[]>(
          "/resource-monitor/alert-events/list",
          {
            method: "POST",
            body: JSON.stringify(input),
          },
        ),
  resolveAlertEvent: (id: number) =>
    hasTauriRuntime()
      ? invoke<void>("resolve_resource_alert_event", { id })
      : devApiFetch<void>(
          `/resource-monitor/alert-events/${encodeURIComponent(id)}/resolve`,
          {
            method: "POST",
          },
        ),
};
